crypto/fipsmodule/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,601 @@
+
+    /*  Number of YMM registers required to store a value */
+    int num_ymm_regs = NUMBER_OF_REGISTERS(red_digits, 256 /* ymm bit size */);
+    /* Capacity of the register set (in qwords = 64-bits) to store a value */
+    int regs_capacity = num_ymm_regs * 4;
+
+    // The following 7 values are in redundant representation and are
+    // to be stored contiguously in storage_aligned as needed by the
+    // function rsaz_mod_exp_x2_ifma256.
+    uint64_t *base1_red, *m1_red, *rr1_red;
+    uint64_t *base2_red, *m2_red, *rr2_red;
+    uint64_t *coeff_red;
+
+    uint64_t *storage = NULL;
+    uint64_t *storage_aligned = NULL;
+    int storage_len_bytes = 7 * regs_capacity * sizeof(uint64_t)
+                           + 64 /* alignment */;
+
+    const uint64_t *exp[2] = {0};
+    uint64_t k0[2] = {0};
+    /* AMM = Almost Montgomery Multiplication */
+    AMM amm = NULL;
+
+    switch (modlen) {
+    case 1024:
+        amm = rsaz_amm52x20_x1_ifma256;
+        break;
+    case 1536:
+        amm = rsaz_amm52x30_x1_ifma256;
+        break;
+    case 2048:
+        amm = rsaz_amm52x40_x1_ifma256;
+        break;
+    default:
+        goto err;
+    }
+
+    storage = (uint64_t *)OPENSSL_malloc(storage_len_bytes);
+    if (storage == NULL)
+        goto err;
+    storage_aligned = (uint64_t *)align_pointer(storage, 64);
+
+    /* Memory layout for red(undant) representations */
+    base1_red = storage_aligned;
+    base2_red = storage_aligned + 1 * regs_capacity;
+    m1_red    = storage_aligned + 2 * regs_capacity;
+    m2_red    = storage_aligned + 3 * regs_capacity;
+    rr1_red   = storage_aligned + 4 * regs_capacity;
+    rr2_red   = storage_aligned + 5 * regs_capacity;
+    coeff_red = storage_aligned + 6 * regs_capacity;
+
+    /* Convert base_i, m_i, rr_i, from regular to 52-bit radix */
+    to_words52(base1_red, regs_capacity, base1, modlen);
+    to_words52(base2_red, regs_capacity, base2, modlen);
+    to_words52(m1_red,    regs_capacity, m1,    modlen);
+    to_words52(m2_red,    regs_capacity, m2,    modlen);
+    to_words52(rr1_red,   regs_capacity, rr1,   modlen);
+    to_words52(rr2_red,   regs_capacity, rr2,   modlen);
+
+    // Based on the definition of n and s above, we have
+    //   R  = 2^n mod m; RR  = R^2 mod m
+    //   R' = 2^s mod m; RR' = R'^2 mod m
+    // To obtain R'^2 from R^2:
+    //   - Let t = AMM(RR, RR) = R^4 / R' mod m
+    //   - Note that R'4 = R^4 * 2^{4*(s-n)} mod m
+    //   - Let k = 4 * (s - n)
+    //   - We have AMM(t, 2^k) = R^4 * 2^{4*(s-n)} / R'^2 mod m
+    //                         = R'^4 / R'^2 mod m
+    //                         = R'^2 mod m
+    
+    OPENSSL_memset(coeff_red, 0, red_digits * sizeof(uint64_t));
+    // coeff_red = 2^k = 1 << bitlen_diff taking into account the
+    // redundant representation in digits of DIGIT_SIZE
+    set_bit(coeff_red, 64 * (int)(bitlen_diff / DIGIT_SIZE) + bitlen_diff % DIGIT_SIZE);
```

Comment:
division by a constant (that's not a power of 2) can produce instruction sequence that's not constant-time. I guess we don't care in this case, but wanted to double check?

Reply:
Ah so true. I don't think we care in this case because `DIGIT_SIZE` is a constant, its non-constant division shouldn't leak anything useful in the context of a timing attack. Would you agree?

Reply:
I agree because the calculation is based on the sizes only which are not secret.